funsor/sum_product.py

```diff
@@ -93,6 +93,98 @@ def partial_sum_product(sum_op, prod_op, factors, eliminate=frozenset(), plates=
     return results
 
 
+def modified_partial_sum_product(sum_op, prod_op, factors,
+                                 eliminate=frozenset(), plate_to_step=dict()):
+    """
+    Generalization of the tensor variable elimination algorithm of
+    :func:`funsor.sum_product.partial_sum_product` to handle markov dimensions
+    in addition to plate dimensions. Markov dimensions are eliminated efficiently
+    using the parallel-scan algorithm in :func:`funsor.sum_product.sequential_sum_product`.
+
+    :param ~funsor.ops.AssociativeOp sum_op: A semiring sum operation.
+    :param ~funsor.ops.AssociativeOp prod_op: A semiring product operation.
+    :param factors: A collection of funsors.
+    :type factors: tuple or list
+    :param frozenset eliminate: A set of free variables to eliminate,
+        including both sum variables and product variable.
+    :param dict plate_to_step: A dict mapping markov dimensions to
+        ``step`` dicts that map previous to current variable name.
+        Plates are passed with an empty ``step``.
+    :return: a list of partially contracted Funsors.
+    :rtype: list
+    """
+    assert callable(sum_op)
+    assert callable(prod_op)
+    assert isinstance(factors, (tuple, list))
+    assert all(isinstance(f, Funsor) for f in factors)
+    assert isinstance(eliminate, frozenset)
+    assert isinstance(plate_to_step, dict)
+    plates = frozenset(plate_to_step.keys())
+    sum_vars = eliminate - plates
+    prod_vars = eliminate.intersection(plates)
+    markov_sum_vars = frozenset()
+    for step in plate_to_step.values():
+        markov_sum_vars |= frozenset(step.keys()) | frozenset(step.values())
+    markov_sum_vars &= sum_vars
+    markov_prod_vars = frozenset(k for k, v in plate_to_step.items() if v and k in eliminate)
+    markov_sum_to_prod = defaultdict(set) 
+    for markov_prod in markov_prod_vars:
```

Comment:
Collects all time dims that a markov var depends on in a set. In line 162 raise intractable error if the set size is not equal 1.